editor/src/components/editor/store/collaborative-editing.ts

```diff
@@ -0,0 +1,363 @@
+      while (!keyIteratorValue.done) {
+        const key = keyIteratorValue.value
+        if (
+          key === change.fullPath ||
+          (change.recursive && key.startsWith(`${change.fullPath}/`))
+        ) {
+          projectContentsMap.delete(key)
+        }
+      }
+      break
+    case 'WRITE_PROJECT_FILE':
+      if (
+        change.projectFile.type === 'TEXT_FILE' &&
+        change.projectFile.fileContents.parsed.type === 'PARSE_SUCCESS'
+      ) {
+        const parsedPart = change.projectFile.fileContents.parsed
+        let collabFile: CollabTextFile
+        if (projectContentsMap.has(change.fullPath)) {
+          collabFile = projectContentsMap.get(change.fullPath)!
+          // TODO Handle not text type
+        } else {
+          collabFile = new Y.Map()
+          collabFile.set('type', 'TEXT_FILE')
+          const topLevelElementsArray = new Y.Array<TopLevelElement>()
+          collabFile.set('topLevelElements', topLevelElementsArray)
+        }
+
+        mergeDoc.transact(() => {
+          if (!projectContentsMap.has(change.fullPath)) {
+            projectContentsMap.set(change.fullPath, collabFile)
+          }
+          synchroniseParseSuccessToCollabFile(parsedPart, collabFile)
+        })
+      }
+      break
+    case 'ENSURE_DIRECTORY_EXISTS':
+      // Not necessary to handle right now.
+      break
+    default:
+      assertNever(change)
+  }
+}
+
+export function updateCollaborativeProjectContents(
+  session: CollaborativeEditingSupportSession,
+  projectChanges: ProjectChanges,
+  filesModifiedByAnotherUser: Array<string>,
+): void {
+  const projectContentsMap = session.projectContents
+  for (const change of projectChanges.fileChanges.collabProjectChanges) {
+    if (!filesModifiedByAnotherUser.includes(change.fullPath)) {
+      applyFileChangeToMap(change, projectContentsMap, session.mergeDoc)
+    }
+  }
+}
+
+export function addHookForProjectChanges(
+  session: CollaborativeEditingSupportSession,
+  dispatch: EditorDispatch,
+): void {
+  session.projectContents.observeDeep((changeEvents) => {
+    // TODO Check that this is the array change before doing anything
+    for (const changeEvent of changeEvents) {
+      switch (changeEvent.path.length) {
+        // This case indicates a change at the base of the entire structure, which
+        // appears to arise at least on first connection to sync up the entire value.
+        case 0: {
+          if (changeEvent instanceof Y.YMapEvent) {
+            updateEntireProjectContents(changeEvent as Y.YMapEvent<any>, dispatch)
+          } else {
+            throw new Error(`Could not treat change event as Y.YMapEvent.`)
+          }
+          break
+        }
+        // Originally thought to be a case that would arise on a new addition of
+        // a file, left here to capture this specific case in case it does show up.
+        case 1: {
+          console.error(`Unhandled path length of 1: ${changeEvent.path}`)
```

Comment:
This should probably be a `throw new Error` then so that it is less likely to be lost in the console

Reply:
True, that's more consistent with the rest of the function as well, corrected in 21cee15c7.